File: frontend/admin/packages/modules/system/src/api/system/base_language.ts
import service from "@liujitcn/kratos-admin-core/request";
import { readonly, shallowRef } from "vue";
import {
	type BaseLanguage,
  type BaseLanguageForm,
	type BaseLanguageService,
  type CreateBaseLanguageRequest,
  type DeleteBaseLanguageRequest,
  type GetBaseLanguageRequest,
  type OptionBaseLanguageRequest,
  type OptionBaseLanguageResponse,
  type PageBaseLanguageRequest,
  type PageBaseLanguageResponse,
  type SetBaseLanguagePrimaryRequest,
  type SetBaseLanguageStatusRequest,
  type UpdateBaseLanguageRequest
} from "@liujitcn/kratos-admin-system/rpc/system/admin/v1/base_language";
import type { Empty } from "@liujitcn/kratos-admin-system/rpc/google/protobuf/empty";

const BASE_LANGUAGE_URL = "/v1/admin/base/language";

/** Admin语言服务。 */
export class BaseLanguageServiceImpl implements BaseLanguageService {
  /** 查询语言选项。 */
  OptionBaseLanguage(request: OptionBaseLanguageRequest = {}): Promise<OptionBaseLanguageResponse> {
    return service<OptionBaseLanguageRequest, OptionBaseLanguageResponse>({
      url: `${BASE_LANGUAGE_URL}/option`,
      method: "get",
      params: request
    });
  }

  /** 查询语言分页列表。 */
  PageBaseLanguage(request: PageBaseLanguageRequest): Promise<PageBaseLanguageResponse> {
    return service<PageBaseLanguageRequest, PageBaseLanguageResponse>({
      url: BASE_LANGUAGE_URL,
      method: "get",
      params: request
    });
  }

  /** 查询语言详情。 */
  GetBaseLanguage(request: GetBaseLanguageRequest): Promise<BaseLanguageForm> {
    return service<GetBaseLanguageRequest, BaseLanguageForm>({
      url: `${BASE_LANGUAGE_URL}/${request.id}`,
      method: "get"
    });
  }

  /** 创建语言。 */
  CreateBaseLanguage(request: CreateBaseLanguageRequest): Promise<Empty> {
    return service<BaseLanguageForm | undefined, Empty>({
      url: BASE_LANGUAGE_URL,
      method: "post",
      data: request.base_language
    });
  }

  /** 更新语言。 */
  UpdateBaseLanguage(request: UpdateBaseLanguageRequest): Promise<Empty> {
    return service<BaseLanguageForm | undefined, Empty>({
      url: `${BASE_LANGUAGE_URL}/${request.base_language?.id ?? ""}`,
      method: "put",
      data: request.base_language
    });
  }

  /** 删除语言。 */
  DeleteBaseLanguage(request: DeleteBaseLanguageRequest): Promise<Empty> {
    return service<DeleteBaseLanguageRequest, Empty>({
      url: `${BASE_LANGUAGE_URL}/${request.id}`,
      method: "delete"
    });
  }

  /** 设置语言启用状态。 */
  SetBaseLanguageStatus(request: SetBaseLanguageStatusRequest): Promise<Empty> {
    return service<SetBaseLanguageStatusRequest, Empty>({
      url: `${BASE_LANGUAGE_URL}/${request.id}/status`,
      method: "put",
      data: request
    });
  }

  /** 设置主语言。 */
  SetBaseLanguagePrimary(request: SetBaseLanguagePrimaryRequest): Promise<Empty> {
    return service<SetBaseLanguagePrimaryRequest, Empty>({
      url: `${BASE_LANGUAGE_URL}/${request.id}/primary`,
      method: "put",
      data: request
    });
  }
}

export const defBaseLanguageService = new BaseLanguageServiceImpl();

const enabledBaseLanguages = shallowRef<BaseLanguage[]>([]);
let enabledBaseLanguagesRequest: Promise<BaseLanguage[]> | undefined;

/** 加载启用语言选项并在当前管理端会话内缓存。 */
export async function loadEnabledBaseLanguages(force = false): Promise<BaseLanguage[]> {
  if (force) {
    enabledBaseLanguagesRequest = undefined;
  }
  if (enabledBaseLanguagesRequest) return enabledBaseLanguagesRequest;
  if (!force && enabledBaseLanguages.value.length) return enabledBaseLanguages.value;

  enabledBaseLanguagesRequest = defBaseLanguageService
    .OptionBaseLanguage({ enabled_only: true })
    .then(response => {
      const languages = response.base_languages ?? [];
      enabledBaseLanguages.value = languages;
      return languages;
    })
    .finally(() => {
      enabledBaseLanguagesRequest = undefined;
    });
  return enabledBaseLanguagesRequest;
}

/** 返回响应式的启用语言选项。 */
export function useEnabledBaseLanguages() {
  return { languages: readonly(enabledBaseLanguages) };
}

/** 返回当前缓存的启用语言选项快照。 */
export function getEnabledBaseLanguages(): BaseLanguage[] {
  return enabledBaseLanguages.value;
}

/** 清除启用语言缓存，使下次使用时重新读取语言管理数据。 */
export function invalidateEnabledBaseLanguages() {
  enabledBaseLanguages.value = [];
  enabledBaseLanguagesRequest = undefined;
}

File: backend/internal/service/system/admin/v1/base_language.go
package admin

import (
	"context"
	"fmt"

	adminv1 "github.com/liujitcn/kratos-admin/backend/api/gen/go/system/admin/v1"
	biz "github.com/liujitcn/kratos-admin/backend/internal/biz/system/admin"
	"github.com/liujitcn/kratos-core/errorsx"

	"github.com/go-kratos/kratos/v3/log"
	"google.golang.org/protobuf/types/known/emptypb"
)

// BaseLanguageService Admin语言管理服务。
type BaseLanguageService struct {
	adminv1.UnimplementedBaseLanguageServiceServer
	baseLanguageCase *biz.BaseLanguageCase
}

// NewBaseLanguageService 创建Admin语言管理服务。
func NewBaseLanguageService(baseLanguageCase *biz.BaseLanguageCase) *BaseLanguageService {
	return &BaseLanguageService{baseLanguageCase: baseLanguageCase}
}

// OptionBaseLanguage 查询启用语言选项。
func (s *BaseLanguageService) OptionBaseLanguage(ctx context.Context, req *adminv1.OptionBaseLanguageRequest) (*adminv1.OptionBaseLanguageResponse, error) {
	res, err := s.baseLanguageCase.OptionBaseLanguage(ctx, req)
	if err != nil {
		log.Error(fmt.Sprintf("OptionBaseLanguage %v", err))
		return nil, errorsx.WrapInternal(err, "查询语言选项失败")
	}
	return res, nil
}

// PageBaseLanguage 查询语言分页列表。
func (s *BaseLanguageService) PageBaseLanguage(ctx context.Context, req *adminv1.PageBaseLanguageRequest) (*adminv1.PageBaseLanguageResponse, error) {
	res, err := s.baseLanguageCase.PageBaseLanguage(ctx, req)
	if err != nil {
		log.Error(fmt.Sprintf("PageBaseLanguage %v", err))
		return nil, errorsx.WrapInternal(err, "查询语言分页列表失败")
	}
	return res, nil
}

// GetBaseLanguage 查询语言详情。
func (s *BaseLanguageService) GetBaseLanguage(ctx context.Context, req *adminv1.GetBaseLanguageRequest) (*adminv1.BaseLanguageForm, error) {
	res, err := s.baseLanguageCase.GetBaseLanguage(ctx, req.GetId())
	if err != nil {
		log.Error(fmt.Sprintf("GetBaseLanguage %v", err))
		return nil, errorsx.WrapInternal(err, "查询语言失败")
	}
	return res, nil
}

// CreateBaseLanguage 创建语言。
func (s *BaseLanguageService) CreateBaseLanguage(ctx context.Context, req *adminv1.CreateBaseLanguageRequest) (*emptypb.Empty, error) {
	err := s.baseLanguageCase.CreateBaseLanguage(ctx, req.GetBaseLanguage())
	if err != nil {
		log.Error(fmt.Sprintf("CreateBaseLanguage %v", err))
		return nil, errorsx.WrapInternal(err, "创建语言失败")
	}
	return new(emptypb.Empty), nil
}

// UpdateBaseLanguage 更新语言。
func (s *BaseLanguageService) UpdateBaseLanguage(ctx context.Context, req *adminv1.UpdateBaseLanguageRequest) (*emptypb.Empty, error) {
	err := s.baseLanguageCase.UpdateBaseLanguage(ctx, req.GetBaseLanguage())
	if err != nil {
		log.Error(fmt.Sprintf("UpdateBaseLanguage %v", err))
		return nil, errorsx.WrapInternal(err, "更新语言失败")
	}
	return new(emptypb.Empty), nil
}

// DeleteBaseLanguage 删除语言。
func (s *BaseLanguageService) DeleteBaseLanguage(ctx context.Context, req *adminv1.DeleteBaseLanguageRequest) (*emptypb.Empty, error) {
	err := s.baseLanguageCase.DeleteBaseLanguage(ctx, req.GetId())
	if err != nil {
		log.Error(fmt.Sprintf("DeleteBaseLanguage %v", err))
		return nil, errorsx.WrapInternal(err, "删除语言失败")
	}
	return new(emptypb.Empty), nil
}

// SetBaseLanguageStatus 设置语言启用状态。
func (s *BaseLanguageService) SetBaseLanguageStatus(ctx context.Context, req *adminv1.SetBaseLanguageStatusRequest) (*emptypb.Empty, error) {
	err := s.baseLanguageCase.SetBaseLanguageStatus(ctx, req)
	if err != nil {
		log.Error(fmt.Sprintf("SetBaseLanguageStatus %v", err))
		return nil, errorsx.WrapInternal(err, "设置语言状态失败")
	}
	return new(emptypb.Empty), nil
}

// SetBaseLanguagePrimary 设置主语言。
func (s *BaseLanguageService) SetBaseLanguagePrimary(ctx context.Context, req *adminv1.SetBaseLanguagePrimaryRequest) (*emptypb.Empty, error) {
	err := s.baseLanguageCase.SetBaseLanguagePrimary(ctx, req)
	if err != nil {
		log.Error(fmt.Sprintf("SetBaseLanguagePrimary %v", err))
		return nil, errorsx.WrapInternal(err, "设置主语言失败")
	}
	return new(emptypb.Empty), nil
}
